llvm/test/CodeGen/AMDGPU/div-rem-by-constant-64.ll

```diff
@@ -0,0 +1,618 @@
+; GFX1030-NEXT:    v_mov_b32_e32 v0, v3
+; GFX1030-NEXT:    v_add_co_u32 v2, s4, v6, v0
+; GFX1030-NEXT:    v_add_co_ci_u32_e64 v3, null, 0, 0, s4
+; GFX1030-NEXT:    v_mad_u64_u32 v[0:1], null, 0xaaaaaaaa, v1, v[2:3]
+; GFX1030-NEXT:    v_alignbit_b32 v0, v1, v0, 1
+; GFX1030-NEXT:    v_lshrrev_b32_e32 v1, 1, v1
+; GFX1030-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %div = udiv i64 %i, 3
+  ret i64 %div
+}
+
+define noundef i64 @udiv64_6(i64 noundef %i)  {
+; GFX9-LABEL: udiv64_6:
+; GFX9:       ; %bb.0: ; %entry
+; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-NEXT:    s_mov_b32 s4, 0xaaaaaaab
+; GFX9-NEXT:    v_mul_hi_u32 v2, v0, s4
+; GFX9-NEXT:    v_mov_b32_e32 v3, 0
+; GFX9-NEXT:    s_mov_b32 s6, 0xaaaaaaaa
+; GFX9-NEXT:    v_mad_u64_u32 v[4:5], s[4:5], v1, s4, v[2:3]
+; GFX9-NEXT:    v_mov_b32_e32 v6, v5
+; GFX9-NEXT:    v_mov_b32_e32 v5, v3
+; GFX9-NEXT:    v_mad_u64_u32 v[2:3], s[4:5], v0, s6, v[4:5]
+; GFX9-NEXT:    v_mov_b32_e32 v0, v3
+; GFX9-NEXT:    v_add_co_u32_e32 v2, vcc, v6, v0
+; GFX9-NEXT:    v_addc_co_u32_e64 v3, s[4:5], 0, 0, vcc
+; GFX9-NEXT:    v_mad_u64_u32 v[0:1], s[4:5], v1, s6, v[2:3]
+; GFX9-NEXT:    v_alignbit_b32 v0, v1, v0, 2
+; GFX9-NEXT:    v_lshrrev_b32_e32 v1, 2, v1
+; GFX9-NEXT:    s_setpc_b64 s[30:31]
+;
+; GFX942-LABEL: udiv64_6:
+; GFX942:       ; %bb.0: ; %entry
+; GFX942-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX942-NEXT:    s_mov_b32 s0, 0xaaaaaaab
+; GFX942-NEXT:    v_mul_hi_u32 v2, v0, s0
+; GFX942-NEXT:    v_mov_b32_e32 v3, 0
+; GFX942-NEXT:    v_mad_u64_u32 v[4:5], s[0:1], v1, s0, v[2:3]
+; GFX942-NEXT:    v_mov_b32_e32 v2, v5
+; GFX942-NEXT:    v_mov_b32_e32 v5, v3
+; GFX942-NEXT:    s_mov_b32 s2, 0xaaaaaaaa
+; GFX942-NEXT:    v_mad_u64_u32 v[4:5], s[0:1], v0, s2, v[4:5]
+; GFX942-NEXT:    v_mov_b32_e32 v4, v5
+; GFX942-NEXT:    v_mov_b32_e32 v5, v3
+; GFX942-NEXT:    v_lshl_add_u64 v[2:3], v[2:3], 0, v[4:5]
+; GFX942-NEXT:    v_mad_u64_u32 v[0:1], s[0:1], v1, s2, v[2:3]
+; GFX942-NEXT:    v_alignbit_b32 v0, v1, v0, 2
+; GFX942-NEXT:    v_lshrrev_b32_e32 v1, 2, v1
+; GFX942-NEXT:    s_setpc_b64 s[30:31]
+;
+; GFX1030-LABEL: udiv64_6:
+; GFX1030:       ; %bb.0: ; %entry
+; GFX1030-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX1030-NEXT:    v_mul_hi_u32 v2, 0xaaaaaaab, v0
+; GFX1030-NEXT:    v_mov_b32_e32 v3, 0
+; GFX1030-NEXT:    v_mad_u64_u32 v[4:5], null, 0xaaaaaaab, v1, v[2:3]
+; GFX1030-NEXT:    v_mov_b32_e32 v6, v5
+; GFX1030-NEXT:    v_mov_b32_e32 v5, v3
+; GFX1030-NEXT:    v_mad_u64_u32 v[2:3], null, 0xaaaaaaaa, v0, v[4:5]
+; GFX1030-NEXT:    v_mov_b32_e32 v0, v3
+; GFX1030-NEXT:    v_add_co_u32 v2, s4, v6, v0
+; GFX1030-NEXT:    v_add_co_ci_u32_e64 v3, null, 0, 0, s4
+; GFX1030-NEXT:    v_mad_u64_u32 v[0:1], null, 0xaaaaaaaa, v1, v[2:3]
+; GFX1030-NEXT:    v_alignbit_b32 v0, v1, v0, 2
+; GFX1030-NEXT:    v_lshrrev_b32_e32 v1, 2, v1
+; GFX1030-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %div = udiv i64 %i, 6
+  ret i64 %div
+}
```

Comment:
Does this need power of 2 tests? What about values around int32_max?

Reply:
Added tests for s/u div/rem for 2, 64, int32 min, int32 max.